beacon-chain/types/block.go

```diff
@@ -128,3 +140,104 @@ func (b *Block) Attestations() []*pb.AttestationRecord {
 func (b *Block) Timestamp() (time.Time, error) {
 	return ptypes.Timestamp(b.data.Timestamp)
 }
+
+// isSlotValid compares the slot to the system clock to determine if the block is valid.
+func (b *Block) isSlotValid() bool {
+	slotDuration := time.Duration(b.SlotNumber()*params.SlotDuration) * time.Second
+	validTimeThreshold := genesisTime.Add(slotDuration)
+
+	return clock.Now().After(validTimeThreshold)
+}
+
+// IsValid is called to decide if an incoming p2p block can be processed into the chain's block trie,
+// it checks time stamp, beacon chain parent block hash. It also checks pow chain reference hash if it's a validator.
+func (b *Block) IsValid(aState *ActiveState, cState *CrystallizedState) bool {
+	_, err := b.Hash()
+	if err != nil {
+		log.Debugf("Could not hash incoming block: %v", err)
+		return false
+	}
+
+	if b.SlotNumber() == 0 {
+		log.Debugf("Can not process a genesis block")
+		return false
+	}
+
+	if !b.isSlotValid() {
+		log.Debugf("slot of block is too high: %d", b.SlotNumber())
+		return false
+	}
+
+	for index, attestation := range b.Attestations() {
+		if !b.isAttestationValid(index, aState, cState) {
+			log.Debugf("attestation invalid: %v", attestation)
+			return false
+		}
+	}
+
+	return true
+}
+
+// isAttestationValid validates an attestation in a block
+func (b *Block) isAttestationValid(attestationIndex int, aState *ActiveState, cState *CrystallizedState) bool {
+	// Validate attestation's slot number has is within range of incoming block number.
+	slotNumber := int(b.SlotNumber())
```

Comment:
Do we need to cast this to int? Can't we leave it as unit64? or cast whatever else to uint64?

Reply:
Hmm I don't know why it was cast here, but seems to work without it. Fixed.

Reply:
Actually, there's a cast because `slotNumber-params.CycleLength` can be a negative number. I'll cast the uint in the if statement where this can happen to make it more obvious.